program/src/main.rs

```diff
@@ -2,99 +2,302 @@
+    println!("vk: {sha256_of_current_vkey_hash:?}");
+
+    let hash_of_public_values = Sha256::digest(&recursive_input.public_values);
+    let mut public_values_buffer = Buffer::from(&recursive_input.public_values);
+
+    sp1_zkvm::io::commit(&recursive_input.genesis_hash);
+    // commit hash of the header being proved
+    sp1_zkvm::io::commit(&recursive_input.current_header.signed_header.header().hash().as_bytes().to_vec());
+
+    let hash_of_proof_vkey : Vec<u8> = public_values_buffer.read();
+
+    let proof_genesis_hash : Vec<u8> = public_values_buffer.read();
+    if proof_genesis_hash != recursive_input.genesis_hash {
+        panic!("invalid genesis");
+    }
+
+    let proven_header_hash : Vec<u8>  = public_values_buffer.read();
+    if proven_header_hash != recursive_input.previous_header.signed_header.header().hash().as_bytes() {
+        panic!("proven hash doesn't match previous header hash");
+    }
+
+    let last_result: bool = public_values_buffer.read();
+    if !last_result {
+        panic!("previous proof invalid");
+    }
+
+    println!("cycle-tracker-start: recursive");
+    match recursive_input.recursive_proof_input {
+        RecursiveProofInput::Sp1 => {
+            let proof_vkey_hash = if *hash_of_proof_vkey == *sha256_of_current_vkey_hash {
+                // previous proof has the same vkey
+                current_vkey_hash_u32
+            } else if *hash_of_proof_vkey == ELF_V1_VK { 
+                // vk hash of previous proof matches hardcoded vk, so verifying that we got passed the
+                // correct key shouldn't be necessary
+                babybears_to_u32(&recursive_input.previous_vkey)
+            } else {
+                panic!("verify sp1: vkey of proof is not one of the allowed vkeys");
+            };
+
+            sp1_zkvm::lib::verify::verify_sp1_proof(&proof_vkey_hash, &hash_of_public_values.into());
+        }
+        RecursiveProofInput::Groth16 { proof, sp1_key } => {
+            let proof_vkey_hash = if *hash_of_proof_vkey == *sha256_of_current_vkey_hash {
+                // previous proof has the same vkey
+                recursive_input.current_vkey
+            } else if *hash_of_proof_vkey == ELF_V1_VK { 
+                // vk hash of previous proof matches hardcoded vk, so verifying that we got passed the
+                // correct key shouldn't be necessary
+                recursive_input.previous_vkey
+            } else {
+                panic!("verify groth16: vkey of proof is not one of the allowed vkeys");
+            };
+
+            // https://docs.rs/sp1-prover/3.0.0-rc1/src/sp1_prover/types.rs.html#56
+            let vkey_digest_bn254 = babybears_to_bn254(&proof_vkey_hash);
+            let vkey_digest_bytes32 = format!("0x{:0>64}", vkey_digest_bn254.as_canonical_biguint().to_str_radix(16));
+            Groth16Verifier::verify(&proof, &recursive_input.public_values, &vkey_digest_bytes32, &sp1_key).expect("groth16 verification failed");
```

Comment:
SP1 key is read from the input, but the program vkey is either hardcoded one of the v1 ELF, or current one; so this seems safe?